sw/device/silicon_creator/manuf/lib/individualize.c

```diff
@@ -75,71 +75,59 @@ static status_t hw_cfg1_enable_knobs_set(const dif_otp_ctrl_t *otp_ctrl) {
+    // boot. Note: cp_device_id and manuf_state are on the same flash info page.
     TRY(flash_ctrl_testutils_info_region_setup_properties(
         flash_state, kFlashInfoFieldCpDeviceId.page,
         kFlashInfoFieldCpDeviceId.bank, kFlashInfoFieldCpDeviceId.partition,
         flash_info_page_0_permissions,
         /*offset=*/NULL));
 
     // Read CpDeviceId from flash info page 0.
-    uint32_t cp_device_id_from_flash[kFlashInfoFieldCpDeviceIdSizeIn32BitWords];
-    uint32_t empty_cp_device_id[kFlashInfoFieldCpDeviceIdSizeIn32BitWords] = {
-        0};
+    uint32_t cp_device_id[kFlashInfoFieldCpDeviceIdSizeIn32BitWords];
     TRY(manuf_flash_info_field_read(flash_state, kFlashInfoFieldCpDeviceId,
-                                    cp_device_id_from_flash,
+                                    cp_device_id,
                                     kFlashInfoFieldCpDeviceIdSizeIn32BitWords));
 
     // Check if CP device ID from flash is empty.
     bool flash_cp_device_id_empty = true;
-    for (size_t i = 0; flash_cp_device_id_empty &&
-                       i < kFlashInfoFieldCpDeviceIdSizeIn32BitWords;
-         ++i) {
-      flash_cp_device_id_empty &= cp_device_id_from_flash[i] == 0;
-    }
-
-    // Check if CP device ID provided is empty (i.e., all ones). This means CP
-    // stage was skipped or already run in the past.
-    bool provided_din_empty = true;
-    for (size_t i = 1;
-         provided_din_empty && i < kFlashInfoFieldCpDeviceIdSizeIn32BitWords;
-         ++i) {
-      provided_din_empty &= device_id[i] == 0;
+    for (size_t i = 0; i < kFlashInfoFieldCpDeviceIdSizeIn32BitWords; ++i) {
+      if (cp_device_id[i] != 0) {
+        flash_cp_device_id_empty = false;
+        break;
+      }
     }
 
-    // If the CP device ID read from flash is non-empty, then it must match the
-    // first 128-bits of device ID provided, unless the first 128-bits are all
-    // ones (in which case we use the CP device ID in flash).
-    //
-    // If the device ID read from flash is empty, we check to ensure the device
-    // ID provided is also not empty, and the CP portion is not all ones. An
-    // empty (all zero) device ID will prevent the keymgr from advancing.
-    if (!flash_cp_device_id_empty) {
-      if (!provided_din_empty) {
-        TRY_CHECK_ARRAYS_EQ(cp_device_id_from_flash, device_id,
-                            kFlashInfoFieldCpDeviceIdSizeIn32BitWords);
+    // On non-silicon targets, we expect the CP device ID from flash to be
+    // empty. In this case we set the HW origin portion of the CP device ID.
+    // Otherwise, we expect the CP device ID to be present and non-zero.
+    if (flash_cp_device_id_empty) {
+      if (kDeviceType != kDeviceSilicon) {
+        cp_device_id[0] = 0x00024001u;
+        memset(&cp_device_id[1], 0,
+               kFlashInfoFieldCpDeviceIdSizeIn32BitWords - 1);
+        memset(&cp_device_id[2], 0,
+               kFlashInfoFieldCpDeviceIdSizeIn32BitWords - 1);
```

Comment:
I don't think this does what you intend.  The size argument to memset is in bytes.  I think this should be something more like:
```
memset(&cp_device_id[1], 0, sizeof(cp_device_id) - sizeof(cp_device_id[0]));
```

However, it's probably even better to memset the entire thing to zero and then fill in the one word that you care about:
```
memset(&cp_device_id, 0, sizeof(cp_device_id));
cp_device_id[0] = 0xWhatEvs;
```

Reply:
ah, yes. nice catch! Fixed.